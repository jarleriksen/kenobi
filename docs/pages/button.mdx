---
name: Button
route: /button
---

import { Playground } from 'docz';
import styled from 'styled-components';
import Button from '../../src/button/src/Button.tsx';
import PurpleButton from '../components/PurpleButton';
import Row from '../components/Row';
import Component from '@reach/component-component';

# Button

### `Button` is a base component of Kenobi. It can contain a label and an icon before or after the label.

## Usage

<Playground>
  <Button>Base button</Button>
</Playground>

## Variations

**Component and Row are not a part of Kenobi**

<Playground>
  <Component initialState={{ isLoading: false }}>
    {({ state, setState }) => (
      <React.Fragment>
        <div style={{ marginBottom: '12px' }}>
          <Button
            onClick={() => setState({ isLoading: !state.isLoading })}
            marginBottom={4}>
            {state.isLoading ? 'Hide' : 'Show'} loading state
          </Button>
        </div>
        <Row>
          <Button marginRight={8} isLoading={state.isLoading}>
            Default
          </Button>
          <Button intent="primary" marginRight={8} isLoading={state.isLoading}>
            Primary
          </Button>
          <Button intent="success" marginRight={8} isLoading={state.isLoading}>
            Success
          </Button>
          <Button intent="warning" marginRight={8} isLoading={state.isLoading}>
            Warning
          </Button>
          <Button intent="danger" marginRight={8} isLoading={state.isLoading}>
            Danger
          </Button>
          <Button marginRight={8} iconBefore="edit" isLoading={state.isLoading}>
            Icon before
          </Button>
          <Button marginRight={8} iconAfter="edit" isLoading={state.isLoading}>
            Icon after
          </Button>
        </Row>
      </React.Fragment>
    )}
  </Component>
</Playground>

[More information on usage of Icons](/icon)

## Sizing

Default size for buttons are 32px.
`Button` will handle its own scaling based on the height.

<Playground>
  <Row>
    <Button height={24} marginRight={8}>
      Height 24
    </Button>
    <Button height={32} marginRight={8}>
      Height 32
    </Button>
    <Button height={40} marginRight={8}>
      Height 40
    </Button>
    <Button height={48} marginRight={8}>
      Height 48
    </Button>
    <Button height={56} marginRight={8}>
      Height 56
    </Button>
  </Row>
</Playground>

## Usage as a link

Due to using Styled-components as our CSS-in-JS library it is really simple to switch out the underlying
HTML element. We just have to pass the prop **isLink** to the Button component.
The button will pass the **as** prop to the Styled element.

This will render the button as an **anchor** element.

<Playground>
  <Button isLink={true} href="/button#button-as-a-link">
    I am a link!
  </Button>
</Playground>

## Extending styles

It's very easy to extend Button styles with styled-components. (Haven't tested any other solutions).

<PurpleButton>I am now Purple!</PurpleButton>

```jsx
import styled from 'styled-components';
import { Button } from 'kenobi-ui';

const PurpleButton = styled(Button)`
  background-color: #551a8b;
`;

<PurpleButton>I am now Purple!</PurpleButton>;
```
