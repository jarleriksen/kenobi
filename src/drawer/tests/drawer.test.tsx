import React from 'react';
import { render } from '@testing-library/react';
import { Drawer } from '..';
import { Button } from '../../button';
import '@testing-library/jest-dom/extend-expect';

describe('Drawer', () => {
  test('should not be in portal when closed', () => {
    const { getByTestId, container, asFragment } = render(
      <Drawer isOpen={false} onRequestClose={() => {}}>
        Content
      </Drawer>,
    );

    const portal = getByTestId('kenobi-portal');
    const drawer: HTMLElement | null = document.querySelector(
      '[data-testid="kenobi-drawer"]',
    );

    expect(portal).not.toContainElement(drawer);
    expect(asFragment()).toMatchSnapshot();
  });

  test('should be in portal when open', () => {
    const { getByTestId, asFragment } = render(
      <Drawer isOpen={true} onRequestClose={() => {}}>
        Content
      </Drawer>,
    );

    const portal = getByTestId('kenobi-portal');
    const drawer: HTMLElement | null = document.querySelector(
      '[data-testid="kenobi-drawer"]',
    );

    expect(portal).toContainElement(drawer);
    expect(asFragment()).toMatchSnapshot();
  });

  test('close button has focus when first opened', () => {
    const { getByTestId, asFragment } = render(
      <Drawer isOpen={true} onRequestClose={() => {}}>
        <div title="insideDrawer" />
      </Drawer>,
    );

    expect(getByTestId('kenobi-drawer-close-button')).toHaveFocus();
    expect(asFragment()).toMatchSnapshot();
  });

  test('can autofocus another element', () => {
    const { getByTestId, container, asFragment } = render(
      <Drawer isOpen={true} onRequestClose={() => {}}>
        <Button data-autofocus>Hello</Button>
      </Drawer>,
    );

    expect(getByTestId('kenobi-button')).toHaveFocus();
    expect(asFragment()).toMatchSnapshot();
  });

  test('contains aria-labelledby', () => {
    const { getByTestId, asFragment } = render(
      <Drawer isOpen={true} ariaLabel="test-drawer" onRequestClose={() => {}}>
        Content
      </Drawer>,
    );

    const drawer = getByTestId('kenobi-drawer');

    expect(drawer).toHaveAttribute('aria-labelledby', 'test-drawer');
    expect(asFragment()).toMatchSnapshot();
  });
});
