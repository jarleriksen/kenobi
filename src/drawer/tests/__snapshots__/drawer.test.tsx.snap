// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Drawer can autofocus another element 1`] = `<DocumentFragment />`;

exports[`Drawer close button has focus when first opened 1`] = `<DocumentFragment />`;

exports[`Drawer contains aria-labelledby 1`] = `<DocumentFragment />`;

exports[`Drawer should be in portal when open 1`] = `<DocumentFragment />`;

exports[`Drawer should not be in portal when closed 1`] = `<DocumentFragment />`;
